app/models/report.rb

```diff
@@ -0,0 +1,177 @@
+class Report
+  attr_accessor :columns, :row_item_type, :rows
+  DEFAULT_COLUMNS = [['Total', :all, :count],
+                     ['Reserved', :reserved, :count],
+                     ['Checked Out', :checked_out, :count],
+                     ['Overdue', :overdue, :count],
+                     ['Returned On Time', :returned_on_time, :count],
+                     ['Returned Overdue', :returned_overdue, :count],
+                     ['User Count', :all, :count, :reserver_id]]
+
+  # Reports are extremely powerful 2D reservation statistics tables
+  # See #build_new for the main constructor method used in the controller.
+  #
+
+  # The Column class contains all the information needed to build the
+  # columns (obviously I should hope!). Simple arrays can be made into
+  # Column objects if their elements are in the right order.
+  #
+  #   name:       human readable name
+  #   filter:     filtering scope to apply to Reservations
+  #   data_type:  how to display the data
+  #   data_field: what field of data to use
+  #   res_set:    ActiveRecord::Relation holding all relevant reservations
+
+  class Column
+    attr_accessor :name, :res_set, :data_type, :filter, :data_field
+    def self.arr_to_col(arr)
+      c = new
+      c.name = arr[0]
+      c.filter = arr[1]
+      c.data_type = arr[2]
+      c.data_field = arr[3]
+      c
+    end
+  end
+
+  # Rows correspond to singular objects like equipment models or users.
+  # You can even construct a row skeleton with one of these objects
+  #
+  #   name:      human readable name
+  #   link_path: clickable link path
+  #   item_id:   the id of the ite
+  #   data:      an array of the row's data
+
+  class Row
+    attr_accessor :name, :link_path, :item_id, :data
+    def self.item_to_row(item)
+      r = Row.new
+      r.link_path = Rails.application.routes.url_helpers.subreport_path(
+        id: item.id, class: item.class.to_s)
```

Comment:
this needs to be `class: item.class.to_s.underscore`, otherwise you get a `ActiveRecord::StatementInvalid` since the `params[:class]` is used to build the column name directly (e.g. `EquipmentModel_id`). It also fits better w/ the Rails style for URLs.